core/src/com/unciv/ui/utils/AskNumberPopup.kt

```diff
@@ -0,0 +1,114 @@
+     * that allows for adding a minus sign, but only when it is the first character. So far so good,
+     * until a user starts typing numbers before an already placed - sign --> crash. Fix that
+     * by disallowing any character being typed in front of a - sign. All is fixed right? Wrong!
+     * Because you now also disallowed writing two minus signs at the same time, copying over a
+     * new number after clamping now disallows overwriting the existing minus sign with a new minus
+     * sign, as there is already a minus sign in the number. Well, no problem, you can just remove
+     * the number before overwriting it with the clamped variant. But now you reset your cursor
+     * position every time you type a character. You might start trying to cache the cursor position
+     * as well, but at that point you're basically rewriting the setText() function, and when I
+     * reached this point I decided to stop. 
+     * 
+     * P.S., if you do decide to go on this quest of adding minus signs, don't forget that 
+     * `"-".toInt()` also crashes, so you need to exclude that before checking to clamp.
+     */
+
+    init {
+        val wrapper = Table()
+        wrapper.add(icon).padRight(10f)
+        wrapper.add(label.toLabel())
+        add(wrapper).colspan(2).row()
+
+        val nameField = TextField(defaultText, skin)
+        nameField.textFieldFilter = TextField.TextFieldFilter.DigitsOnlyFilter()
+        
+        fun clampInBounds(int: Int): Int {
+            if (bounds[0] != null && bounds[0]!! > int) {
+                return bounds[0]!!
+            }
+            if (bounds[1] != null && bounds[1]!! < int)
+                return bounds[1]!!
+            return int
+        } 
+        
+        nameField.onChange {
+            if (nameField.text.isNotEmpty()) {
+                nameField.text = clampInBounds(nameField.text.toInt()).toString()
+            }
+        }
+        
+        val centerTable = Table(skin)
+        val sortedAmountValues = amountButtons.sorted()
+        
+        fun addValueButton(value: Int) {
+            centerTable.add(
+                Button(
+                    if (value > 0) "+$value".toLabel()
+                    else value.toLabel(),
+                    skin
+                ).apply {
+                    onClick {
+                        nameField.text = clampInBounds(nameField.text.toInt() + value).toString()
+                    }
+                }
+            ).pad(5f)
+        }
+        
+        for (value in sortedAmountValues.reversed()) {
+            addValueButton(-value)
+        }
+
+        centerTable.add(nameField).growX().pad(10f)
+        
+        add(centerTable).colspan(2).row()
+
+        for (value in sortedAmountValues) {
+            addValueButton(value)
+        }
+        
+        val errorLabel = errorText.toLabel()
+        errorLabel.color = Color.RED
+
+        addOKButton(automaticallyCloseOnPress = false) {
```

Comment:
I'd model the usecase more like other frameworks as open fun validate() = true or somesuch. If that returns false, block OK-action (now we need a sound - possibly quite flatulent-sounding :trumpet: )
_or_ the text entry box gets a kind of Type enum that controls filter, listener, and OK validation without exposing the intrinsics of it, and this class would only add the buttons.